source/extensions/http/cache/file_system_http_cache/file_system_http_cache.cc

```diff
@@ -388,6 +405,78 @@ void FileSystemHttpCache::trackFileRemoved(uint64_t file_size) {
   stats_.size_bytes_.set(size_bytes_);
 }
 
+void FileSystemHttpCache::maybeEvict() {
+  uint64_t size_to_evict = 0;
+  uint64_t count_to_evict = 0;
+  if (config().has_max_cache_size_bytes()) {
+    // capture a value from the atomic because we're going to use it twice and don't want our
+    // value to change in between.
+    uint64_t seen_size = size_bytes_;
+    size_to_evict = seen_size > config().max_cache_size_bytes().value()
+                        ? seen_size - config().max_cache_size_bytes().value()
+                        : 0;
+  }
+  if (config().has_max_cache_entry_count()) {
+    // capture a value from the atomic because we're going to use it twice and don't want our
+    // value to change in between.
+    uint64_t seen_count = size_count_;
+    count_to_evict = seen_count > config().max_cache_entry_count().value()
+                         ? seen_count - config().max_cache_entry_count().value()
+                         : 0;
+  }
+  if (size_to_evict == 0 && count_to_evict == 0) {
+    return;
+  }
+
+  auto os_sys_calls = Api::OsSysCallsSingleton::get();
+  uint64_t size = 0;
+  uint64_t count = 0;
+  uint64_t proposed_size_evicted = 0;
+  struct ProposedEviction {
+    std::string name_;
+    uint64_t size_;
+    struct timespec last_touch_;
+    bool operator<(const ProposedEviction& other) const { return last_touch_ < other.last_touch_; }
+  };
+  std::multiset<ProposedEviction> proposed_evictions;
+
+  for (const Filesystem::DirectoryEntry& entry : Filesystem::Directory(std::string{cachePath()})) {
```

Comment:
it seems like an interface bug that Filesystem::Directory doesn't allow a string_view. Can we make its interface be a string_view and avoid having to stringify things at every call-site?

This would be a separate PR; I'm not asking for changes on this one for this.

Reply:
Agree that would be nicer. Or at least making it take a std::string rather than a std::string&, because it just copies it into its own string anyway, so it's having to perform *two* copies here unless the optimizer reasons it away.
(Taking a string_view would reduce it to one copy in this context, but so would taking a std::string and moving it. For convenience *and* performance, having both std::string and absl::string_view constructors would be optimal, because e.g. absl::StrCat(...) passed in as a string_view would perform an additional copy, vs. absl::StrCat(...) passed in as a std::string would require only a move.)